### DISCUSSION TITLE
Pre DOM changes async callback and custom fetch headers

### DISCUSSION THREAD

**Discussion by Yahav** [PaperStrike/Pjax#175] [Ideas]
In order for this package to be used in most projects powered by JS frameworks, it would need to be a little more customizable. 
Two examples are below:

1. Providing an option to add custom headers to the requests:
In my project this will be used to provide the XCSRF header required to validate the CSRF token.
Without it the request will be denied .

2.  Providing an option for async callback after the response has been received and before the DOM changes applied:
This will be useful in order to prepare and init whatever required for the new page before it is loaded to DOM.

Without those two, it will be real hard to have a real life usage for this package IMHO.

**Comment by Yahav** [comment]
As for 2, i'm not sure that async was what i really meant, which is a function that will need to be complete execution before the dom change is starting..

**Comment by PaperStrike** [comment]
Thanks for your suggestions. But I haven't understood everything, yet.

For 1, I wonder how did you add your custom headers for a normal navigation like clicking an `<a>` link (without Pjax). The [default trigger](https://github.com/PaperStrike/Pjax#defaulttrigger) is all for simulating standard behaviour, so it will unlikely to be supported if the feature you requested isn't in the standard.
If you are not using the default trigger, [`load`](https://github.com/PaperStrike/Pjax#load) and [`weakLoad`](https://github.com/PaperStrike/Pjax#weakload) do support using a [Request](https://developer.mozilla.org/en-US/docs/Web/API/Request) as the target, you can add your headers there.

**Comment by PaperStrike** [comment]
Released in 2.2.0 [hooks](https://github.com/PaperStrike/Pjax/tree/v2.2.0#hooks). Thank you!
